src/CommitmentsV2/SFA.DAS.CommitmentsV2/Application/Commands/AddCohort/AddCohortHandler.cs

```diff
@@ -0,0 +1,212 @@
+
+namespace SFA.DAS.CommitmentsV2.Application.Commands.AddCohort
+{
+    public class AddCohortHandler : IRequestHandler<AddCohortCommand, AddCohortResponse>
+    {
+        private readonly Lazy<ProviderCommitmentsDbContext> _dbContext;
+        private readonly IHashingService _hashingService;
+        private readonly ILogger<AddCohortHandler> _logger;
+
+        public AddCohortHandler(Lazy<ProviderCommitmentsDbContext> dbContext, IHashingService hashingService, ILogger<AddCohortHandler> logger)
+        {
+            _dbContext = dbContext;
+            _hashingService = hashingService;
+            _logger = logger;
+        }
+
+        public async Task<AddCohortResponse> Handle(AddCohortCommand command, CancellationToken cancellationToken)
+        {
+            var db = _dbContext.Value;
+
+            var commitment = await SaveCohortWithLogging(db, command, cancellationToken);
+
+            var response = new AddCohortResponse
+            {
+                Id = commitment.Id,
+                Reference = commitment.Reference,
+                DraftApprenticeshipId = commitment.Apprenticeship.Single().Id
+            };
+
+            return response;
+        }
+
+        private async Task<Commitment> SaveCohortWithLogging(ProviderCommitmentsDbContext db, AddCohortCommand command, CancellationToken cancellationToken)
+        {
+            Exception exception = null;
+            Commitment result = null;
+            try
+            {
+                result = await SaveCohort(db, command, cancellationToken);
+                return result;
+            }
+            catch (Exception ex)
+            {
+                exception = ex;
+                throw;
+            }
+            finally
+            {
+                if (exception == null)
+                {
+                    _logger.LogInformation($"Saved-commitment Provider: {command.ProviderId} Account-Legal-Entity:{command.AccountLegalEntityId} Reservation-Id:{command.ReservationId} Commitment-Id:{result?.Id} Apprenticeship:{result?.Apprenticeship?.First()?.Id}");
+                }
+                else
+                {
+                    _logger.LogError(exception, $"Saving-commitment provider: {command.ProviderId} account-legal-entity:{command.AccountLegalEntityId} reservation-id:{command.ReservationId}");
+                }
+            }
+        }
+
+        private async Task<Commitment> SaveCohort(ProviderCommitmentsDbContext db, AddCohortCommand command, CancellationToken cancellationToken)
+        {
+            var commitment = await AddCommitment(db, command, cancellationToken);
+            var apprentice = AddDraftApprenticeship(command);
+
+            commitment.Apprenticeship.Add(apprentice);
+
+            await db.SaveChangesAsync(cancellationToken);
+
+            commitment.Reference = _hashingService.HashValue(commitment.Id);
+
+            await db.SaveChangesAsync(cancellationToken);
+
```

Comment:
We need to do the save twice, as you saw (we need the hash). We can have a discussion about whether sequences would be useful, but bare in mind this would itself require a round-trip to the database.